In [26]:
import pandas as pd
from pathlib import Path

In [27]:
ROOT = Path.cwd().resolve().parent
DATA = ROOT / "Data" 
GDP  = DATA / "gdp-per-capita-worldbank" / "gdp-per-capita-worldbank.csv"
LIFE = DATA / "happiness-cantril-ladder" / "happiness-cantril-ladder.csv"

In [28]:
df_gdp = pd.read_csv(GDP)
df_life = pd.read_csv(LIFE)

In [29]:
df_gdp.head()

Entity Code  Year  \
0  Afghanistan  AFG  2000   
1  Afghanistan  AFG  2001   
2  Afghanistan  AFG  2002   
3  Afghanistan  AFG  2003   
4  Afghanistan  AFG  2004   

   GDP per capita, PPP (constant 2021 international $)  \
0                                          1617.8264     
1                                          1454.1108     
2                                          1774.3087     
3                                          1815.9282     
4                                          1776.9182     

  World region according to OWID  
0                            NaN  
1                            NaN  
2                            NaN  
3                            NaN  
4                            NaN

In [30]:
df_life.head()

Entity Code  Year  Cantril ladder score
0  Afghanistan  AFG  2011                 4.258
1  Afghanistan  AFG  2012                 4.040
2  Afghanistan  AFG  2014                 3.575
3  Afghanistan  AFG  2015                 3.360
4  Afghanistan  AFG  2016                 3.794

In [31]:
df_gdp.isnull().sum()

Entity                                                    0
Code                                                    435
Year                                                      0
GDP per capita, PPP (constant 2021 international $)      75
World region according to OWID                         7039
dtype: int64

In [32]:
df_life.isnull().sum()

Entity                    0
Code                    130
Year                      0
Cantril ladder score      0
dtype: int64

In [33]:
len(df_life['Entity'].unique())

178

In [34]:
len(df_gdp['Entity'].unique())

285

In [35]:
285-178

107

Based on the information shown, we can conclude that the merge should be done using `Entity` rather than `Code`, because some entities do not have a code in the dataset. If we merged by `code`, we would lose information.

Additionally, the GDP dataset contains more entities—107 more, to be precise. Therefore, merging by code would result in losing at least 107 countries in the final dataset.

In [36]:
df_gdp.columns

Index(['Entity', 'Code', 'Year',
       'GDP per capita, PPP (constant 2021 international $)',
       'World region according to OWID'],
      dtype='object')

In [37]:
df_life = df_life[['Entity', 'Year', 'Code', 'Cantril ladder score']]
df_gdp = df_gdp[['Entity', 'Year', 'GDP per capita, PPP (constant 2021 international $)']]

In [38]:
df_life_gdp = pd.merge(left=df_life, right=df_gdp ,on=['Entity','Year'], how='inner')

In [39]:
df_life_gdp

Entity  Year Code  Cantril ladder score  \
0     Afghanistan  2011  AFG                 4.258   
1     Afghanistan  2012  AFG                 4.040   
2     Afghanistan  2014  AFG                 3.575   
3     Afghanistan  2015  AFG                 3.360   
4     Afghanistan  2016  AFG                 3.794   
...           ...   ...  ...                   ...   
1970     Zimbabwe  2020  ZWE                 3.145   
1971     Zimbabwe  2021  ZWE                 2.995   
1972     Zimbabwe  2022  ZWE                 3.204   
1973     Zimbabwe  2023  ZWE                 3.341   
1974     Zimbabwe  2024  ZWE                 3.396   

      GDP per capita, PPP (constant 2021 international $)  
0                                             2757.0525    
1                                             2985.3190    
2                                             3017.9426    
3                                             2967.6921    
4                                             2958.7854    
...                                                 ...    
1970                                          2987.2683    
1971                                          3184.7847    
1972                                          3323.1184    
1973                                          3442.2488    
1974                                          3450.1296    

[1975 rows x 5 columns]

In [40]:
len(df_life_gdp['Entity'].unique())

168

We can see that we have 168 entities because the merge also takes the year into account, and it is possible that some countries did not match for certain years.

In [41]:
Mex = df_life_gdp[df_life_gdp['Entity']=='Mexico']
Mex

Entity  Year Code  Cantril ladder score  \
1173  Mexico  2011  MEX                 6.740   
1174  Mexico  2012  MEX                 7.088   
1175  Mexico  2014  MEX                 7.187   
1176  Mexico  2015  MEX                 6.778   
1177  Mexico  2016  MEX                 6.578   
1178  Mexico  2017  MEX                 6.488   
1179  Mexico  2018  MEX                 6.595   
1180  Mexico  2019  MEX                 6.465   
1181  Mexico  2020  MEX                 6.317   
1182  Mexico  2021  MEX                 6.128   
1183  Mexico  2022  MEX                 6.330   
1184  Mexico  2023  MEX                 6.678   
1185  Mexico  2024  MEX                 6.979   

      GDP per capita, PPP (constant 2021 international $)  
1173                                          20457.600    
1174                                          20898.936    
1175                                          21069.766    
1176                                          21408.928    
1177                                          21578.264    
1178                                          21777.523    
1179                                          21997.773    
1180                                          21704.186    
1181                                          19728.482    
1182                                          20782.588    
1183                                          21392.111    
1184                                          21905.178    
1185                                          22033.275

In [42]:
df_life_gdp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1975 entries, 0 to 1974
Data columns (total 5 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   Entity                                               1975 non-null   object 
 1   Year                                                 1975 non-null   int64  
 2   Code                                                 1923 non-null   object 
 3   Cantril ladder score                                 1975 non-null   float64
 4   GDP per capita, PPP (constant 2021 international $)  1972 non-null   float64
dtypes: float64(2), int64(1), object(2)
memory usage: 77.3+ KB


Now we have to make the datasets of the entities for 2021, 2022, 2023, 2024 the amount of entities may change in the datasets, as we can see in Mexico there is no record for 2013. That may be the case for some entities between 2021 and 2024.

In [43]:
DATA_YEAR = DATA / "data_for_year"

In [ ]:
years = [2021, 2022, 2023, 2024]

for year in years:
    (
        df_life_gdp[df_life_gdp["Year"] == year]
        .to_csv(DATA_YEAR / f"lifesat_gdp_{year}.csv", index=False)
    )